### PR TITLE
fix: DBTP-1138 Add progress tracking for deployments

### PR DESCRIPTION
**Pull request by samuele-mattiuzzo**

### CODE REVIEW

File: image_builder/commands/deploy.py [hunk 1]
```diff
@@ -94,10 +98,19 @@ def deploy(send_notifications):
...
+            progress.current_phase_success()
             deploy_status_msg = "deployed to"
 
+        extras = {
+            "repository_name": codebase_repository,
+            "revision_commit": commit_hash,
+            "repository_url": f"https://github.com/{codebase_repository}",
+        }
+        notify.post_build_progress(progress, None, extras)
+
```
Comment (antroy-madetech) [inline image_builder/commands/deploy.py:113]: Looks like there should be tests around this new behaviour in the "test_deploy.py" file.

File: test/image_builder/test_notify.py [hunk 2]
```diff
@@ -28,6 +28,11 @@ def setUp(self):
         self.codebase.revision.get_repository_url.return_value = (
             "https://github.com/org/repo"
         )
+        self.extras = {
+            "repository_name": "org/repo",
+            "revision_commit": "commit-sha",
+            "repository_url": "https://github.com/org/repo",
+        }
```
Comment (antroy-madetech) [inline test/image_builder/test_notify.py:35]: Would be good to have these as different to the values in the codebase example so it's obvious they are coming from different places.

File: test/image_builder/test_notify.py [hunk 3]
```diff
@@ -84,6 +89,36 @@ def test_sending_progress_updates(self, webclient, time):
...
         )
 
+    def test_sending_progress_updates_no_codebase(self, webclient, time):
+        notify = Notify(self.codebase)
+        progress = Progress()
+        notify.post_build_progress(progress, None, self.extras)
+        notify.post_build_progress(progress, None, self.extras)
+        notify.slack.chat_postMessage.assert_called_with(
+            channel="channel-id",
+            blocks=ANY,
```
Comment (antroy-madetech) [inline test/image_builder/test_notify.py:99]: Should be checking the value of these "blocks" parts. I appreciate they weren't tested in the original test you copied, but I think both should be checking the values in all of the blocks entries rather than just ANY, as the data is now relevant.

File: test/image_builder/test_notify.py [hunk 4]
```diff
@@ -84,6 +89,36 @@ def test_sending_progress_updates(self, webclient, time):
...
+        notify.slack.chat_update.assert_called_with(
+            channel="channel-id",
+            blocks=ANY,
+            ts="updated-message",
+            text="Building: org/repo@commit-sha",
+            unfurl_links=False,
+            unfurl_media=False,
+        )
+
     def test_sending_progress_updates_when_notifications_off(self, webclient, time):
```
Comment (antroy-madetech) [inline test/image_builder/test_notify.py:122]: Also need tests for when both parameters (codebase and extras) are present, and when neither are (codebase=None and no extras)

**Review by antroy-madetech — COMMENTED** [review]
More tests required.

File: test/image_builder/test_notify.py (hunk 3 above)
Comment (samuele-mattiuzzo) [reply test/image_builder/test_notify.py:99]: There's also a few more extra things I'm looking at. We're instantiating `Notify(codebase)` but its `__init__` takes an argument to set notifications on/off. So we're using the existence of an object to enable notifications and I think it's a side effect that isn't great nor clear. More so, `Notify` has a codebase attribute that doesn't seem to be needed anywhere.

File: image_builder/notify.py [hunk 5]
```diff
@@ -33,19 +33,33 @@ def __init__(self, send_notifications: bool = True):
...
+        if text_blocks is not None:
+            return text_blocks
+
+        if codebase is not None:
+            return {
+                "repository_name": codebase.revision.get_repository_name(),
+                "revision_commit": codebase.revision.commit,
+                "repository_url": codebase.revision.get_repository_url(),
+            }
+        return {}
```
Comment (WillGibson) [inline image_builder/notify.py:46]: What would be the expected behaviour `if text_blocks is not None and codebase is not None`?

Feels like it should maybe throw an exception?

Comment (samuele-mattiuzzo) [reply image_builder/notify.py:46]: I think so https://github.com/uktrade/ci-image-builder/pull/72#discussion_r1679635539 see my previous answer to Ant

**Comment by samuele-mattiuzzo** [comment]
@WillGibson @antroy-madetech would you mind taking another look? I've refactored a fair bit now, decoupling `Notify` from `Codebase`

File: test/image_builder/test_notify.py (hunk 3 above)
Comment (samuele-mattiuzzo) [reply test/image_builder/test_notify.py:99]: @antroy-madetech @WillGibson one thing I think it's worth doing that I've just thought about is to add a validation step that checks the structure of the data that's now passed in to `post_build_progress`

File: image_builder/notify.py [hunk 6]
```diff
@@ -33,19 +31,19 @@ def __init__(self, send_notifications: bool = True):
             except KeyError as e:
                 raise ValueError(f"{e} environment variable must be set")
 
-    def post_build_progress(self, progress: Progress, codebase: Codebase):
+    def post_build_progress(self, progress: Progress, text_blocks=None):
```
Comment (WillGibson) [inline image_builder/notify.py:34]: I think we need a test for when `text_blocks` is `None`.

File: image_builder/commands/deploy.py [hunk 7]
```diff
@@ -49,6 +50,7 @@ def deploy(send_notifications):
         repository = get_image_repository_url()
         Docker.login(repository.split("/")[0])
         timestamp = get_deployment_reference(repository, tag)
+        progress = Progress()
```
Comment (antroy-madetech) [inline image_builder/commands/deploy.py:53]: Not used for a while, can probably shift down to around line 67

File: image_builder/notify.py (hunk 6 above)
Comment (WillGibson) [reply image_builder/notify.py:48]: If `text_blocks` is never meant to be anything other than a `Dict` of a set structure, could we:

* create a class with those three properties
* put any required validation/defaulting into the constructor of the class
* remove the validation above and change this line to...

```suggestion
    def post_build_progress(self, progress: Progress, text_blocks: WhateverWeCallTheClass):
```

I'm also wondering if `text_blocks` is the best name for that variable. Maybe `progress_data`?

Comment (samuele-mattiuzzo) [reply image_builder/notify.py:48]: so we remove the tight coupling between `Notify` and `Codebase` to add a new coupling between `Notify` and `WhateverWeCallTheClass`

What I've done, In my mind at least, is one step forward towards a generic `Notify` object that takes pre-formatted slack blocks and posts them (akin to what @antroy-madetech suggests)

In truth, `ci-image-builder`'s notification object should be a `NotificationFormatter` object that pings into the `Notify` object, but it's a bit of scope creep

**Review by antroy-madetech — APPROVED** [review]
Small minor point.

File: image_builder/notify.py (hunk 6 above)
Comment (WillGibson) [reply image_builder/notify.py:48]: Feels like there are further iterations to do on this code, but we probably have more pressing concerns right now.

**Review by WillGibson — APPROVED** [review]
